### PR TITLE
Overhaul Tests + Errors for Button (deprecated) and ButtonV1

### PR DESCRIPTION
**Pull request by lawrencewin**
### Platforms Impacted
- [ ] iOS
- [ ] macOS
- [x] win32 (Office)
- [x] windows
- [ ] android

### Description of changes

Previously a part of #2423, split to be more managable.

Currently, for most developers, it is difficult to figure out what the root issue is when a FURN E2E test fails because of opaque error messages. The existing E2E tests are also in need of some clean up in terms of code and PageObject design. 

This commit is part of a larger effort to address the above. Here, I've updated error messages, use new BasePage methods, and did general cleanup for the following tests:
- Button (deprecated)
- ButtonV1

#### Removal of Selectors

One big change to the tests across the board is the removal of Selector enums within tests. Previously, Selectors were used to choose which element was to be clicked, get an attribute from, etc. However, because many tests required different selectors from the base `ComponentSelector` in BasePage.ts to select test elements, this meant that different Page Objects ended up implementing the same methods (`sendKey()`, `click()`, and other attribute testing methods). 

In removing selectors, we can re-use methods for doing most things a test requires implemented in BasePage.ts and pass WebdriverIO elements directly (see [`compareAttribute()`](https://github.com/microsoft/fluentui-react-native/blob/3500d25b492a0fba0df3ab478b09d2775b78e452/apps/E2E/src/common/BasePage.ts#L80-L93),[ `click()`](https://github.com/microsoft/fluentui-react-native/blob/3500d25b492a0fba0df3ab478b09d2775b78e452/apps/E2E/src/common/BasePage.ts#L127-L130), and [`sendKeys()`](https://github.com/microsoft/fluentui-react-native/blob/3500d25b492a0fba0df3ab478b09d2775b78e452/apps/E2E/src/common/BasePage.ts#L132-L134)). This way, future Page Objects remain light and only implement getters or other methods needed for the specific test.

This doesn't mean that Selectors can't be used. Taking the [FocusZone test](https://github.com/microsoft/fluentui-react-native/blob/3500d25b492a0fba0df3ab478b09d2775b78e452/apps/E2E/src/FocusZone/pages/FocusZonePageObject.ts#L18-L28), the large amount of elements used within the test let us use a Selector Enum to pick out specific elements to pass in. But for most cases, we can rely on a small amount of getters and the reusable BasePage methods for creating a test.

### Verification

Tests pass in the CI and locally. 

image

### Pull request checklist

This PR has considered (when applicable):
- [x] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: apps/E2E/src/ButtonExperimental/specs/ButtonExperimental.spec.win.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ describe('Experimental Button Accessibility Testing', async () => {
     await expect(await ButtonExperimentalPageObject.didAssertPopup()).toBeFalsy(ButtonExperimentalPageObject.ERRORMESSAGE_ASSERT);
   });
 
-  it('Experimental Button - Set accessibilityLabel', async () => {
+  it('Set "accessibilityLabel" prop. Validate "accessibilityLabel" value propagates to "Name" element attribute.', async () => {
```
Comment (ksiler) [inline apps/E2E/src/ButtonExperimental/specs/ButtonExperimental.spec.win.ts:40]: Not sure I like adding "propagates to "XXX" element attribute." as part of the test descriptions. I suppose it's helpful in letting the user know which native property it's looking for 🤔

File: apps/E2E/src/Button/specs/Button.spec.win.ts [hunk 2]
```diff
@@ -30,18 +26,35 @@ describe('Button Accessibility Testing', () => {
     await ButtonPageObject.scrollToTestElement();
   });
 
-  it('Button - Validate accessibilityRole is correct', async () => {
-    await expect(await ButtonPageObject.getAccessibilityRole()).toEqual(BUTTON_A11Y_ROLE);
+  it('Validate "accessibilityRole" value is correct, propagates to "ControlType" element attribute.', async () => {
```
Comment (ksiler) [inline apps/E2E/src/Button/specs/Button.spec.win.ts:29]: I think it'd be useful to mention what control type we're expecting so I'd reword this as "Validate "accessibilityRole" value defaults to "Button" ControlType." or something similar.

File: apps/E2E/src/ButtonExperimental/specs/ButtonExperimental.spec.win.ts (hunk 1 above)
Comment (lawrencewin) [reply apps/E2E/src/ButtonExperimental/specs/ButtonExperimental.spec.win.ts:40]: The "propagates to "XXX" element attribute" is more verbose, but is a lot more clear on what the test actually does as opposed to stating "is correct". I guess it's better to be over-descriptive than vague.

Comment (ksiler) [reply apps/E2E/src/ButtonExperimental/specs/ButtonExperimental.spec.win.ts:40]: I can agree with that.

File: apps/E2E/src/Button/specs/Button.spec.win.ts [hunk 3]
```diff
@@ -30,18 +26,35 @@ describe('Button Accessibility Testing', () => {
...
+        BUTTON_ACCESSIBILITY_LABEL_DEPRECATED,
+      ),
+    ).toBeTrue();
+
     await expect(await ButtonPageObject.didAssertPopup()).toBeFalsy(ButtonPageObject.ERRORMESSAGE_ASSERT);
   });
 
-  it('Button - Do not set accessibilityLabel -> Default to Button label', async () => {
-    await expect(await ButtonPageObject.getAccessibilityLabel(ComponentSelector.Secondary)).toEqual(BUTTON_TEST_COMPONENT_LABEL_DEPRECATED);
+  it('Do NOT set "accessibilityLabel" prop. Validate "accessibilityLabel" value defaults to the button label.', async () => {
```
Comment (ksiler) [inline apps/E2E/src/Button/specs/Button.spec.win.ts:49]: Should we say "Validate the "Name" element attribute value defaults to the button label"?

Comment (lawrencewin) [reply apps/E2E/src/Button/specs/Button.spec.win.ts:49]: Yes we should.

**Review by ksiler — APPROVED** [review]

File: apps/E2E/src/ButtonExperimental/pages/ButtonExperimentalPageObject.ts [hunk 4]
```diff
@@ -10,14 +10,9 @@ class ButtonExperimentalPageObject extends BasePage {
   /******************************************************************/
   /**************** UI Element Interaction Methods ******************/
   /******************************************************************/
-  async didOnClickCallbackFire(): Promise<boolean> {
+  async didOnClickCallbackFire(errorMsg?: string): Promise<boolean> {
```
Comment (samuelfreiberg) [inline apps/E2E/src/ButtonExperimental/pages/ButtonExperimentalPageObject.ts:13]: Same as above - no need for optionality

File: apps/E2E/src/Button/pages/ButtonPageObject.ts [hunk 5]
```diff
@@ -6,41 +6,17 @@ import {
...
   /******************************************************************/
-  async didOnClickCallbackFire(): Promise<boolean> {
-    const callbackText = await By(BUTTON_ON_PRESS_DEPRECATED);
-    await browser.waitUntil(async () => await callbackText.isDisplayed(), {
-      timeout: this.waitForUiEvent,
-      timeoutMsg: 'The OnClick callback did not fire.',
-      interval: 1000,
-    });
 
+  async didOnClickCallbackFire(errorMsg?: string): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (lawrencewin) [reply apps/E2E/src/Button/pages/ButtonPageObject.ts]: Removed the optional for this and the v1 page object

File: apps/E2E/src/ButtonExperimental/pages/ButtonExperimentalPageObject.ts (hunk 4 above)
Comment (lawrencewin) [reply apps/E2E/src/ButtonExperimental/pages/ButtonExperimentalPageObject.ts]: Done

File: apps/E2E/src/ButtonLegacy/pages/ButtonLegacyPageObject.ts [hunk 6]
```diff
@@ -6,41 +6,17 @@ import {
...
-    const callbackText = await By(BUTTON_ON_PRESS_DEPRECATED);
-    await browser.waitUntil(async () => await callbackText.isDisplayed(), {
-      timeout: this.waitForUiEvent,
-      timeoutMsg: 'The OnClick callback did not fire.',
-      interval: 1000,
-    });
 
+  async didOnClickCallbackFire(errorMsg: string): Promise<boolean> {
+    const callbackText = await this._callbackText;
+    await this.waitForCondition(async () => await callbackText.isDisplayed(), errorMsg ?? 'The onClick callback did not fire.');
```
Comment (samuelfreiberg) [inline apps/E2E/src/ButtonLegacy/pages/ButtonLegacyPageObject.ts:16]: No need for the `?? 'The onClick callback did not fire.'`.

Now that errorMsg is not optional, we know it'll be passed it

File: apps/E2E/src/ButtonV1/pages/ButtonV1PageObject.ts [hunk 7]
```diff
@@ -10,14 +10,9 @@ class ButtonV1PageObject extends BasePage {
...
-  async didOnClickCallbackFire(): Promise<boolean> {
+  async didOnClickCallbackFire(errorMsg: string): Promise<boolean> {
     const callbackText = await By(BUTTON_ON_PRESS);
-    await browser.waitUntil(async () => await callbackText.isDisplayed(), {
-      timeout: this.waitForUiEvent,
-      timeoutMsg: 'The OnClick callback did not fire.',
-      interval: 1000,
-    });
-
+    await this.waitForCondition(async () => await callbackText.isDisplayed(), errorMsg ?? 'The onClick callback did not fire.');
```
Comment (samuelfreiberg) [inline apps/E2E/src/ButtonV1/pages/ButtonV1PageObject.ts:15]: No need for `?? 'The onClick callback did not fire.'`

File: apps/E2E/src/ButtonV1/specs/ButtonV1.spec.win.ts [hunk 8]
```diff
@@ -64,31 +64,35 @@ describe('ButtonV1 Functional Testing', async () => {
...
-    await expect(await ButtonV1PageObject.didOnClickCallbackFire()).toBeTruthy(
-      `The primary button failed to fire an onClick callback with a mouse click.`,
-    );
+    await expect(
+      await ButtonV1PageObject.didOnClickCallbackFire(`The primary button failed to fire an onClick callback with a mouse click.`),
+    ).toBeTruthy();
+
     await expect(await ButtonV1PageObject.didAssertPopup()).toBeFalsy(ButtonV1PageObject.ERRORMESSAGE_ASSERT);
 
     await ButtonV1PageObject.click(ButtonV1PageObject._primaryComponent); // Reset Button State
```
Comment (samuelfreiberg) [inline apps/E2E/src/ButtonV1/specs/ButtonV1.spec.win.ts:75]: What exactly does this reset? Should this be in an `afterEach()` hook?

File: apps/E2E/src/ButtonLegacy/specs/ButtonLegacy.spec.win.ts [hunk 9]
```diff
@@ -27,22 +26,39 @@ describe('Button Legacy Accessibility Testing', () => {
     await ButtonLegacyPageObject.scrollToTestElement();
   });
 
-  it('Button - Validate accessibilityRole is correct', async () => {
-    await expect(await ButtonLegacyPageObject.getAccessibilityRole()).toEqual(BUTTON_A11Y_ROLE);
+  it('Validate "accessibilityRole" defaults to Button "ControlType" element attribute.', async () => {
```
Comment (samuelfreiberg) [inline apps/E2E/src/ButtonLegacy/specs/ButtonLegacy.spec.win.ts:29]: Let's remove the "ControlType" part

'Validate "accessibilityRole" defaults to ControlType.Button'

File: apps/E2E/src/ButtonV1/pages/ButtonV1PageObject.ts [hunk 10]
```diff
@@ -10,17 +10,23 @@ class ButtonV1PageObject extends BasePage {
...
-      interval: 1000,
-    });
-
+  async waitForOnClickCallbackToFire(errorMsg: string): Promise<boolean> {
+    const callbackText = await this._callbackText;
+    await this.waitForCondition(async () => await callbackText.isDisplayed(), errorMsg);
     return await callbackText.isDisplayed();
   }
 
+  async resetTest(): Promise<void> {
```
Comment (samuelfreiberg) [inline apps/E2E/src/ButtonV1/pages/ButtonV1PageObject.ts:19]: Shouldn't this also apply to ButtonLegacy?

Comment (lawrencewin) [reply apps/E2E/src/ButtonV1/pages/ButtonV1PageObject.ts:19]: It should. Added the same hook in its test as well.

**Review by samuelfreiberg — APPROVED** [review]
